knime_extension/src/nodes/io.py

```diff
@@ -0,0 +1,289 @@
+        self.geo_col = knut.column_exists_or_preset(
+            configure_context, self.geo_col, input_schema, knut.is_geo
+        )
+        return None
+
+    def execute(self, exec_context: knext.ExecutionContext, input_1):
+        exec_context.set_progress(
+            0.4, "Writing file (This might take a while without progress changes)"
+        )
+        gdf = gp.GeoDataFrame(input_1.to_pandas(), geometry=self.geo_col)
+        if self.dataformat == "Shapefile":
+            fileurl = knut.ensure_file_extension(self.data_url, ".shp")
+            gdf.to_file(fileurl)
+        else:
+            fileurl = knut.ensure_file_extension(self.data_url, ".geojson")
+            gdf.to_file(fileurl, driver="GeoJSON")
+        return None
+
+
+############################################
+# GeoPackage Reader
+############################################
+@knext.node(
+    name="GeoPackage Reader",
+    node_type=knext.NodeType.SOURCE,
+    icon_path=__NODE_ICON_PATH + "GeoPackageReader.png",
+    category=__category,
+    after="",
+)
+@knext.output_table(
+    name="Geodata table",
+    description="Geodata from the input file path.",
+)
+@knext.output_table(
+    name="Geodata Layer",
+    description="Layer information from the input file path.",
+)
+@knut.geo_node_description(
+    short_description="Read GeoPackage layer",
+    description="""This node reads [Geopackage,](https://www.geopackage.org/) GeoDatabase(GDB) files. 
+
+You can specify the layer to read. If the layer is empty or wrong, the node will read the first layer. 
+You can also enter the number of the layer to read starting at 0. The node will output the names of all layers as 
+second output table, which can be used to revise the name of the target layer.
+
+Examples of standard local file paths are *C:\\KNIMEworkspace\\test.gpkg* for Windows and
+*/KNIMEworkspace/test.gpkg* for Linux. The node can also load resources directly from a web URL.
+
+**Note:** For larger files the node progress might not change for a time until the file is successfully read.
+    """,
+    references={
+        "Reading Spatial Data": "https://geopandas.org/en/stable/docs/user_guide/io.html",
+        "Read file": "https://geopandas.org/en/stable/docs/reference/api/geopandas.read_file.html",
+    },
+)
+class GeoPackageReaderNode:
+    data_url = knext.StringParameter(
+        "Input file path",
+        "The file path for reading data.",
+        "",
+    )
+
+    data_layer = knext.StringParameter(
+        "Input layer name or order for reading",
+        "The layer name in the multiple-layer data.",
+        "",
+    )
+
+    def configure(self, configure_context):
+        # TODO Create combined schema
+        return None
+
+    def execute(self, exec_context: knext.ExecutionContext):
+        exec_context.set_progress(
+            0.4, "Reading file (This might take a while without progress changes)"
+        )
+        layerlist = fiona.listlayers(self.data_url)
```

Comment:
To improve the startup speed of geospatial Python processes, it definitely helps to reduce the number of imports at the top level. E.g. here `fiona` is only used in this execute method, so you could write `import fiona` before this line in the execute method. This defers the import to the point when it is actually needed.

If a module gets imported multiple times in the same Python process, only the first import actually does something, the next one will notice that the module is already loaded.